mountpoint-s3-crt/src/io/stream.rs

```diff
@@ -0,0 +1,192 @@
+//! AWS input streams.
+
+use crate::common::{allocator::Allocator, error::Error};
+use crate::CrtError;
+use mountpoint_s3_crt_sys::*;
+use std::marker::PhantomData;
+use std::ptr::NonNull;
+
+/// Status of an [InputStream].
+#[derive(Debug)]
+pub struct StreamStatus {
+    is_valid: bool,
+    is_end_of_stream: bool,
+}
+
+impl From<aws_stream_status> for StreamStatus {
+    fn from(status: aws_stream_status) -> Self {
+        Self {
+            is_valid: status.is_valid,
+            is_end_of_stream: status.is_end_of_stream,
+        }
+    }
+}
+
+impl From<StreamStatus> for aws_stream_status {
+    fn from(status: StreamStatus) -> Self {
+        Self {
+            is_valid: status.is_valid,
+            is_end_of_stream: status.is_end_of_stream,
+        }
+    }
+}
+
+/// Specifies where to seek from in an [InputStream].
+#[derive(Debug)]
+pub enum SeekBasis {
+    /// Seek from the beginning of the stream.
+    Begin,
+    /// Seek from the end of the stream.
+    End,
+}
+
+impl From<aws_stream_seek_basis> for SeekBasis {
+    fn from(value: aws_stream_seek_basis) -> Self {
+        match value {
+            aws_stream_seek_basis::AWS_SSB_BEGIN => Self::Begin,
+            aws_stream_seek_basis::AWS_SSB_END => Self::End,
+            _ => panic!("invalid stream seek basis: {value:?}"),
+        }
+    }
+}
+
+impl From<SeekBasis> for aws_stream_seek_basis {
+    fn from(value: SeekBasis) -> Self {
+        match value {
+            SeekBasis::Begin => aws_stream_seek_basis::AWS_SSB_BEGIN,
+            SeekBasis::End => aws_stream_seek_basis::AWS_SSB_END,
+        }
+    }
+}
+
+/// An [InputStream] is a way to read bytes. They can be obtained either from CRT functions,
+/// or by creating a new one based on a Rust type that implements the [GenericInputStream] trait.
+#[derive(Debug)]
+pub struct InputStream<'a> {
+    /// The inner aws_input_stream
+    pub(crate) inner: NonNull<aws_input_stream>,
+
+    /// Phantom data to keep the lifetimes correct, for example, if this stream is created from an
+    /// aws_byte_cursor that has some lifetime.
+    _phantom: PhantomData<&'a [u8]>,
+}
+
+impl<'a> InputStream<'a> {
+    /// Create a new [InputStream] from a slice. The slice is not copied, and so the resulting
+    /// [InputStream] cannot outlive the slice (enforced by a lifetime restriction on the [InputStream].
+    pub fn new_from_slice(allocator: &Allocator, buffer: &'a [u8]) -> Result<Self, Error> {
+        let cursor = aws_byte_cursor {
+            len: buffer.len(),
+            ptr: buffer.as_ptr() as *mut u8,
+        };
+
+        // SAFETY: allocator is a valid aws_allocator. `Self` has a lifetime of 'a, so Rust
+        // will ensure that the return value from this function doesn't out live the buffer.
+        // We need to make sure in things that consume InputStream that we don't accidentally
```

Comment:
could you please describe a wrong usage of `InputStream`? is it possible to make an attempt to read from a destroyed `buffer` from rust code using the interface defined below?

(no questions regarding `put_object_single` impl, as I believe we rely on the fact that CRT won't access the stream after invoking `finish_callback`, nor will it access the `aws_http_message` holding the pointer to it)